File: Program_4/4(b).c
#include <stdio.h>
#include <time.h>

void fibonachi_by_iterative(int n) {
    int x = 0;
    int y = 1;
    n = n - 2;
    // printf("%d %d ", x, y);
    for (int i = 0; i < n; i++) {
        if (i % 2 == 0) {
            x = x + y;
            // printf("%d ", x);
        } else {
            y = y + x;
            // printf("%d ", y);
        }
    }
}

int main() {
    int sizes[] = {100, 1000, 5000, 10000};
    int num_sizes = 4;
    int repetitions = 10000; 

    for (int i = 0; i < num_sizes; i++) {
        int n = sizes[i];

        clock_t start = clock();
        for (int j = 0; j < repetitions; j++) {
            fibonachi_by_iterative(n);
        }
        clock_t end = clock();

        double time = ((double)(end - start)) / CLOCKS_PER_SEC / repetitions;
        printf("for size=%d time=%f\n", n, time);  
    }

    return 0;
}

File: Program_4/4(b).py
import time
import matplotlib.pyplot as plt

def fibonacchi_by_iterative(n):
    x=0
    y=1
    n=n-2
    #print(x,y,end=" ")
    for i in range(n):
        if(i%2==0):
            x=x+y
            #print(x,end=" ")
        else:
            y=y+x
            #print(y,end=" ")

sizes = [10, 100, 500, 1000]
times = []

for n in sizes:
    start = time.perf_counter()
    fibonacchi_by_iterative(n)
    end = time.perf_counter()
    elapsed = end - start
    times.append(elapsed)
    print(f"For input size {n}, time taken = {elapsed:.6f} sec")

plt.plot(sizes, times, marker='o', label="Recursive Sequence Generator")
plt.xlabel("Input Size (n)")
plt.ylabel("Time Taken (seconds)")
plt.title("Iterative Fibonacci Sequence Generator Performance")
plt.legend()
plt.grid(True)
plt.show()
